awscli/utils.py

```diff
@@ -539,3 +540,31 @@ def __init__(self):
 
     def visit_shape(self, shape):
         self.visited.append(shape)
+
+
+def add_component_to_user_agent_extra(session, component):
+    if session.user_agent_extra and not session.user_agent_extra.endswith(" "):
+        session.user_agent_extra += " "
+    session.user_agent_extra += f"{component.to_string()}"
+
+
+def add_metadata_component_to_user_agent_extra(session, name, value=None):
+    add_component_to_user_agent_extra(
+        session,
+        UserAgentComponent("md", name, value)
+    )
+
+
+_LINEAGE_COMMAND_METADATA_REGEX = re.compile(r'md\/command#[\w\.]*')
+
+
+def add_command_lineage_to_user_agent_extra(session, lineage):
+    # Find existing command metadata and append the last item in the lineage to
+    # it
+    if comp := _LINEAGE_COMMAND_METADATA_REGEX.search(session.user_agent_extra):
```

Comment:
Looking at this now, I plan take a deeper look into this to see if we actually need this logic. This is definitely a bit messy and it would be nice if we could avoid it all together.

Reply:
Circling back on this... So I think this pull request made the chance of the session having the command component less significant: https://github.com/aws/aws-cli/pull/7399. Specifically, the log will eagerly delegate to the subcommand's `__call__` instead of updating the user agent.

That being said there are instances such as `s3 rb --force` where a command is called directly: https://github.com/aws/aws-cli/blob/4613a2ef56e4df28325e0fd9d1c9187ae2477717/awscli/customizations/s3/subcommands.py#L913 but that ends up in unexpected updates to the command component (e.g. `md/command#s3.rb.rm` on the DeleteObject calls).

I think with this knowledge, we can probably just update it to check if `md/command` is defined and only add the command if it is not in the user agent. That should help simplify this logic.